File: m3_adapter/gvd/surface.py
# -*- coding: utf-8 -*-
"""surface.py — 2D 地面可通行层(surface places)。

纯函数, 与 3D GVD / 物体层解耦。设计: 输入一张 2D 可通行 bool 栅格(地面投影),
聚类成 surface place 区域(连通分量), 返回区域 + 门口相邻边。nav_frontend(每帧
局部反应式)与后端(累积全局)共用同一份聚类算法 —— 只是喂的栅格 scope 不同。

坐标: 全部在导航平面(ROS map/odom 系, z-up 的水平面 x-y)。不涉及 vxw y-up。
"""
from __future__ import annotations

import numpy as np
from scipy import ndimage


def cluster_surface_places(walkable, origin_xy, res, min_cells=15,
                           door_bridge_m=0.9):
    """把 2D 可通行 bool 栅格聚成 surface place 区域。

    Args:
        walkable: (ny, nx) bool, True = 可走地面格。
        origin_xy: (x0, y0) 栅格 (0,0) 格中心对应的世界米坐标原点(左下角)。
        res: 米/格。
        min_cells: 小于此格数的区域丢弃(碎片)。
        door_bridge_m: 判定两区域"经门相邻"的膨胀桥接距离(米)。

    Returns:
        regions: list[dict] {id, centroid_xy, area_m2, cell_count, label_id}
        edges:   list[(i, j)] 区域间可通行相邻(门口)
        label_img: (ny,nx) int, 每格的区域 label(0=非可走/被丢弃)
    """
    lab, n = ndimage.label(walkable, structure=np.ones((3, 3), np.uint8))
    regions = []
    keep_label = {}
    for r in range(1, n + 1):
        cells = np.argwhere(lab == r)          # (k, 2) = (row=y, col=x)
        if len(cells) < min_cells:
            lab[lab == r] = 0                  # 丢弃碎片
            continue
        cen = cells.mean(0)                    # (cy, cx) in cells
        wx = origin_xy[0] + (cen[1] + 0.5) * res
        wy = origin_xy[1] + (cen[0] + 0.5) * res
        rid = len(regions)
        keep_label[r] = rid
        regions.append({
            "id": rid,
            "centroid_xy": (float(wx), float(wy)),
            "area_m2": float(len(cells) * res * res),
            "cell_count": int(len(cells)),
            "label_id": int(r),
        })

    # 门口相邻: 膨胀每个区域 ~半个门宽, 若与另一区域相交 => 经门连通
    edges = []
    bridge = max(int(round(door_bridge_m / res / 2)), 1)
    struct = ndimage.generate_binary_structure(2, 2)
    masks = {}
    for reg in regions:
        m = (lab == reg["label_id"])
        masks[reg["id"]] = ndimage.binary_dilation(m, struct, iterations=bridge)
    ids = [reg["id"] for reg in regions]
    for ai in range(len(ids)):
        for bi in range(ai + 1, len(ids)):
            a, b = ids[ai], ids[bi]
            if (masks[a] & (lab == regions[bi]["label_id"])).any():
                edges.append((a, b))
    return regions, edges, lab


def occupancy_from_counts(surf_hits, obst_hits, surf_min=1, obst_min=2):
    """counts -> (walkable bool, costmap int8)。
    costmap: -1 未知 / 0 可走 / 100 障碍(nav_msgs/OccupancyGrid 约定)。"""
    walkable = (surf_hits >= surf_min) & (obst_hits < obst_min)
    blocked = obst_hits >= obst_min
    cost = np.full(surf_hits.shape, -1, dtype=np.int8)
    cost[walkable] = 0
    cost[blocked] = 100
    return walkable, cost

File: semantic_map_core/src/surface.cpp
// surface.cpp — 见 surface.hpp。移植自 surface.py(scipy ndimage.label + binary_dilation)。
#include "semantic_map_core/surface.hpp"

#include <cmath>
#include <queue>

namespace smc {

void occupancy_from_counts(const int* surf_hits, const int* obst_hits, int n,
                           int surf_min, int obst_min,
                           std::vector<uint8_t>& walkable, std::vector<int8_t>& cost) {
  walkable.assign(n, 0);
  cost.assign(n, -1);
  for (int i = 0; i < n; ++i) {
    bool w = (surf_hits[i] >= surf_min) && (obst_hits[i] < obst_min);
    bool b = (obst_hits[i] >= obst_min);
    walkable[i] = w ? 1 : 0;
    if (w) cost[i] = 0;
    if (b) cost[i] = 100;  // 与 Python 顺序一致: 先写 walkable=0, 再 blocked=100
  }
}

// 8 连通、光栅顺序首遇编号(复刻 scipy.ndimage.label 的连续编号): 行优先扫描,
// 遇到未标记的 walkable 格就 BFS 整个分量, label 递增。
static int label_cc(const uint8_t* walkable, int ny, int nx, std::vector<int>& lab) {
  lab.assign((size_t)ny * nx, 0);
  int next = 0;
  std::queue<int> q;
  for (int r = 0; r < ny; ++r) {
    for (int c = 0; c < nx; ++c) {
      int idx = r * nx + c;
      if (!walkable[idx] || lab[idx]) continue;
      ++next;
      lab[idx] = next;
      q.push(idx);
      while (!q.empty()) {
        int cur = q.front(); q.pop();
        int cr = cur / nx, cc = cur % nx;
        for (int dr = -1; dr <= 1; ++dr)
          for (int dc = -1; dc <= 1; ++dc) {
            if (!dr && !dc) continue;
            int nr = cr + dr, ncc = cc + dc;
            if (nr < 0 || nr >= ny || ncc < 0 || ncc >= nx) continue;
            int ni = nr * nx + ncc;
            if (walkable[ni] && !lab[ni]) { lab[ni] = next; q.push(ni); }
          }
      }
    }
  }
  return next;
}

// mask 就地 8 连通膨胀一步后 &passable。
static void dilate1_and(std::vector<uint8_t>& mask, const std::vector<uint8_t>& passable,
                        int ny, int nx, std::vector<uint8_t>& out) {
  out.assign((size_t)ny * nx, 0);
  for (int r = 0; r < ny; ++r)
    for (int c = 0; c < nx; ++c) {
      int idx = r * nx + c;
      bool any = false;
      for (int dr = -1; dr <= 1 && !any; ++dr)
        for (int dc = -1; dc <= 1 && !any; ++dc) {
          int nr = r + dr, ncc = c + dc;
          if (nr < 0 || nr >= ny || ncc < 0 || ncc >= nx) continue;
          if (mask[nr * nx + ncc]) any = true;
        }
      out[idx] = (any && passable[idx]) ? 1 : 0;
    }
}

SurfaceResult cluster_surface_places(const uint8_t* walkable, int ny, int nx,
                                     double origin_x, double origin_y, double res,
                                     int min_cells, double door_bridge_m,
                                     const uint8_t* blocked) {
  SurfaceResult out;
  std::vector<int> lab;
  int n = label_cc(walkable, ny, nx, lab);

  // 按 label 1..n 顺序: 小分量丢弃(lab=0), 否则建区域(id=保留计数, label_id=r)。
  for (int r = 1; r <= n; ++r) {
    long cnt = 0; double sr = 0, sc = 0;
    for (long i = 0; i < (long)ny * nx; ++i)
      if (lab[i] == r) { ++cnt; sr += i / nx; sc += i % nx; }
    if (cnt < min_cells) {
      for (long i = 0; i < (long)ny * nx; ++i) if (lab[i] == r) lab[i] = 0;
      continue;
    }
    double cen_r = sr / cnt, cen_c = sc / cnt;
    double wx = origin_x + (cen_c + 0.5) * res;
    double wy = origin_y + (cen_r + 0.5) * res;
    SurfRegion reg;
    reg.id = (int)out.regions.size();
    reg.cx = wx; reg.cy = wy;
    reg.area_m2 = (double)cnt * res * res;
    reg.cell_count = (int)cnt;
    reg.label_id = r;
    out.regions.push_back(reg);
  }

  // 门口相邻: 每区域从 (lab==label_id) 起, 测地膨胀 bridge 步(8连通, 每步 &passable),
  // 若够到另一区域的格 => 连边。
  int bridge = (int)std::rint(door_bridge_m / res);  // Python round=银行家舍入
  if (bridge < 1) bridge = 1;
  std::vector<uint8_t> passable((size_t)ny * nx, 1);
  if (blocked)
    for (long i = 0; i < (long)ny * nx; ++i) passable[i] = blocked[i] ? 0 : 1;

  std::vector<std::vector<uint8_t>> masks(out.regions.size());
  for (size_t ri = 0; ri < out.regions.size(); ++ri) {
    int lid = out.regions[ri].label_id;
    std::vector<uint8_t> m((size_t)ny * nx, 0), tmp;
    for (long i = 0; i < (long)ny * nx; ++i) m[i] = (lab[i] == lid) ? 1 : 0;
    for (int s = 0; s < bridge; ++s) { dilate1_and(m, passable, ny, nx, tmp); m.swap(tmp); }
    masks[ri] = std::move(m);
  }
  for (size_t ai = 0; ai < out.regions.size(); ++ai)
    for (size_t bi = ai + 1; bi < out.regions.size(); ++bi) {
      int lidb = out.regions[bi].label_id;
      bool hit = false;
      for (long i = 0; i < (long)ny * nx && !hit; ++i)
        if (masks[ai][i] && lab[i] == lidb) hit = true;
      if (hit) out.edges.emplace_back(out.regions[ai].id, out.regions[bi].id);
    }

  out.label_img = std::move(lab);
  return out;
}

}  // namespace smc
